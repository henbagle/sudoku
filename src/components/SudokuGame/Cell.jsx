import React from 'react';

const Cell = ({cell, onFocus, tryInputNumber, color}) => {

    const inputClasses = `h-12 w-12 ${calculateBorderStyle(cell.coord)} ${calculateBgColor(cell, color)} 
                        focus:ring focus:ring-inset 
                        text-2xl text-center ${cell.editable ? "": "font-semibold bg-white"}`

    const displayValue = ((cell.value === null) ? "" : cell.value)

    return (
        <div>
            <CellPossibleValues values={cell.possibleValues} />
            <input 
                type="text" 
                value={displayValue}
                onFocus={onFocus}
                onChange={tryInputNumber}
                className={inputClasses} 
                disabled={!cell.editable}>
            </input>
        </div>
    );
}

function CellPossibleValues(props)
{
    if(props.values && props.values.length > 0)
    {
        return (<div className="absolute px-1 py-0.5 text-xs w-12 truncate pointer-events-none"
                aria-label="possible values">
                    {props.values.join(" ")}
                </div>)
    }
    else return null
}

function calculateBgColor(cell, color)
{
    if(!cell.isValid || color == "red")
    {
        return "bg-red-100"
    }
    else return "focus:bg-blue-50";
}

function calculateBorderStyle(coord)
{
    const INITIAL_BORDER = "border border-gray-700";
    const T = " border-t-2";
    const L = " border-l-2";
    const R = " border-r-2";
    const B = " border-b-2";

    let outputStyle = INITIAL_BORDER;
    const subgridX = coord.c % 3; // Position within local 3x3 grid
    const subgridY = coord.r % 3;

    switch(subgridX)
    {
        case 0:
            outputStyle += L;
            break;
        case 2:
            outputStyle += R;
            break;
        default: break;
    }

    switch(subgridY)
    {
        case 0:
            outputStyle += T;
            break;
        case 2:
            outputStyle += B;
            break;
        default: break;
    }

    return outputStyle;
}

export default Cell;