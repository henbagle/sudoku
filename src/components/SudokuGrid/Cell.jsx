import React from 'react';

const Cell = ({detVal, editable, coords, onFocus, tryInputNumber, color}) => {

    const inputClasses = `h-12 w-12 ${calculateBorderStyle(coords)} ${calculateBgColor(color)} focus:ring focus:ring-inset focus:bg-blue-50 
                        text-2xl text-center ${editable ? "": "font-semibold bg-white"}`
    const value = ((detVal === 0) ? "" : detVal)

    return ( 
        <input 
            type="text" 
            value={value}
            onFocus={onFocus}
            onChange={tryInputNumber}
            className={inputClasses} 
            disabled={!editable}>
        </input>
    );
}

function calculateBgColor(color)
{
    switch (color) {
        case "row": return "bg-red-100";
        case "col": return "bg-green-100";
        case "grid": return "bg-yellow-100";
        default: return "";
    }
}

function calculateBorderStyle(coords)
{
    const INITIAL_BORDER = "border border-gray-700";
    const T = " border-t-2";
    const L = " border-l-2";
    const R = " border-r-2";
    const B = " border-b-2";

    let outputStyle = INITIAL_BORDER;
    const subgridX = coords.c % 3; // Position within local 3x3 grid
    const subgridY = coords.r % 3;

    switch(subgridX)
    {
        case 0:
            outputStyle += L;
            break;
        case 2:
            outputStyle += R;
            break;
        default: break;
    }

    switch(subgridY)
    {
        case 0:
            outputStyle += T;
            break;
        case 2:
            outputStyle += B;
            break;
        default: break;
    }

    return outputStyle;
}
 
export default Cell;