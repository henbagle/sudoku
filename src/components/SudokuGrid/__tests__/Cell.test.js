import React from "react";
import renderer from "react-test-renderer";
import Cell from "../Cell.jsx";

const emptyCell = {
    detVal: 0,
    editable: true,
    coords: { r: 1, c: 1 },
    onFocus: () => {},
    tryInputNumber: () => {},
};
const defaultCell = {
    detVal: 1,
    editable: false,
    coords: { r: 2, c: 2 },
    onFocus: () => {},
    tryInputNumber: () => {},
};

test("Cell gets rendered", () => {
    const emptyComponent = renderer.create(<Cell {...emptyCell} />);
    const defaultComponent = renderer.create(<Cell {...defaultCell} />);

    let tree1 = emptyComponent.toJSON();
    expect(tree1).toMatchSnapshot();

    let tree2 = defaultComponent.toJSON();
    expect(tree2).toMatchSnapshot();
});
