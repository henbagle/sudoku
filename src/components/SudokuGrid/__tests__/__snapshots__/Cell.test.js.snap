// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell gets rendered 1`] = `
<input
  className="h-12 w-12 border border-gray-700  focus:ring focus:ring-inset focus:bg-blue-50 
                        text-2xl text-center "
  disabled={false}
  onChange={[Function]}
  onFocus={[Function]}
  type="text"
  value=""
/>
`;

exports[`Cell gets rendered 2`] = `
<input
  className="h-12 w-12 border border-gray-700 border-r-2 border-b-2  focus:ring focus:ring-inset focus:bg-blue-50 
                        text-2xl text-center font-semibold bg-white"
  disabled={true}
  onChange={[Function]}
  onFocus={[Function]}
  type="text"
  value={1}
/>
`;
